oeps/oep-38-Data-Modeling.rst

```diff
@@ -0,0 +1,442 @@
+should make sense to most people familiar with the domain, without
+needing much clarifying documentation.
+
+Prioritize ease of analysis
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For every hour of engineer-time spent creating a data model, several
+thousand more are spent using the model and analyzing its data. Taking
+the time to consider how the data will be used, and thoughtfully design
+our data representation, can potentially save many hours during the
+analysis phase.
+
+This is especially true since changing data models, once they are in
+use, can be time-consuming. A data model change in a core application
+can affect many other systems. Small changes to a data model could
+possibly cause hours or days of work for different teams throughout the
+organization:
+
+-  Data engineering and Analytics to update their workflows
+
+-  BI teams to update reports that utilize this data,
+
+-  Devops to deploy migrations to large tables, with the risk of
+      potential downtime.
+
+Due to the potentially high cost of changing a model, it pays to get it
+as correct as possible the first time.
+
+Think about the person who will analyze this data later and the person
+who manages the system day to day. (That person could possibly be future
+you!) Try to do as much careful design up front to make your quality of
+life better later.
+
+History is best (but events are a close second)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Accuracy is best when history about a change is captured at the database
+level when the change happens. Changes in data relating to business
+continuity should be stored historically where the change is made,
+preferably in the same system as the system of record. This allows us to
+reconstruct the data at different points with much greater reliability.
+
+If, for some reason, it is not possible to create history at the time of
+entry (for example, if we expect the table to become too large or the
+writes to be too frequent), We recommend that an event be issued in its
+place.
+
+Fire events for CUD operations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We highly recommend firing events for any create, update or delete (CUD)
+operations that happen within the OpenedX ecosystem. The cost of firing
+and storing these events is low, and will allow teams in the future to
+reconstruct the state of the application at various times.
+
+Data Modeling Standards
+-----------------------
+
+All models in the edX ecosystem should have:
+
+-  A Primary Key
+
+   -  Should be named: “id”
+
+   -  Use numeric based identifiers
+
+   -  A key size large enough so the keys will not run out for a very
+         long time.
+
+   -  Do not use composite based primary keys. Use a primary key column
+
+-  Updated and created timestamps
+
+   -  Use UTC time.
+
+   -  Created date should be named: “created_date”
+
+   -  Updated date should be named: “modified_date”
+
+-  Use a numeric foreign key(s)
+
+   -  Data should be joined on other foreign keys
```

Comment:
Thanks for pointing this out I've cleaned up the wording here. The idea is that you should join on other Primary Keys instead of joining on things such as username.